### PR TITLE
fix: compatibility with typing_extensions 4.13 and type statement

### LINKED ISSUE #12473: typing-extensions 4.13.0 possibly breaks ORM declarative codepaths that are deprecated in 2.0
### Describe the bug

Defining a simple custom type with the `type` syntax (es. `type MyType = Literal["test1", "test2"]`) and using it within a declarative model within a Mapped (e.g. `Mapped[MyType]`) fails with typing-extensions `4.13.0`

I raised the same issue on typing_extensions and it seems something that needs to be fixed on SQLAlchemy side (just like pydantic had to fix it)
 https://github.com/python/typing_extensions/issues/560

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

_No response_

### SQLAlchemy Version in Use

2.0.39

### DBAPI (i.e. the database driver)

psycopg(3) (but shouldn't matter)

### Database Vendor and Major Version

Postgres 13 (but shouldn't matter)

### Python Version

3.12.9

### Operating system

Linux (Ubuntu)

### To Reproduce

```python
from typing import Literal

from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

type MyType = Literal["test1", "test2"]


class Base(DeclarativeBase):
    pass


class MyModel(Base):
    __tablename__ = "my_model"

    id: Mapped[int] = mapped_column(primary_key=True)
    my_type: Mapped[MyType]
```

### Error

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "../lib/python3.12/site-packages/sqlalchemy/orm/decl_api.py", line 842, in __init_subclass__
    _as_declarative(cls._sa_registry, cls, cls.__dict__)
  File "../lib/python3.12/site-packages/sqlalchemy/orm/decl_base.py", line 244, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "../lib/python3.12/site-packages/sqlalchemy/orm/decl_base.py", line 325, in setup_mapping
    return _ClassScanMapperConfig(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "../lib/python3.12/site-packages/sqlalchemy/orm/decl_base.py", line 572, in __init__
    self._extract_mappable_attributes()
  File "../lib/python3.12/site-packages/sqlalchemy/orm/decl_base.py", line 1568, in _extract_mappable_attributes
    value.declarative_scan(
  File "../lib/python3.12/site-packages/sqlalchemy/orm/properties.py", line 707, in declarative_scan
    self._init_column_for_annotation(
  File "../lib/python3.12/site-packages/sqlalchemy/orm/properties.py", line 871, in _init_column_for_annotation
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Could not locate SQLAlchemy Core type for Python type MyType inside the 'my_type' attribute Mapped annotation
``` 

### Additional context

It doesn't matter whether you add the type in the type_annotation_map within Base.

### PR DESCRIPTION
**Pull request by Daraan**
`typing_extensions` 4.13 has a backport `typing_extensions.TypeAliasType` which differs from `typing.TypeAliasType` which is created though the `type` statement.

`isinstance` which only checks for the `typing_extensions` variant will return `False` which can lead to errors.

(future) TODOs:
- [x] add tests for `typing.TypeAliasType`

Fixes: #12473

### CODE REVIEW

**Comment by zzzeek** [comment]
oh we have one

**Comment by Daraan** [comment]
I came here from https://github.com/python/typing_extensions/issues/560 which I am looking into

VSCode was to eager in creating the PR :D

**Comment by Giuzzilla** [comment]
I opened an issue: https://github.com/sqlalchemy/sqlalchemy/issues/12473#issue-2950415406

**Comment by CaselIT** [comment]
thanks for the pr. we will likely take it over since there is also a test to change, also there may be further cases were we are doing some assumptions

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 8861a5acfb8e81663413ff144b41abf64779b6fd of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 8861a5acfb8e81663413ff144b41abf64779b6fd: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5825

**Review by sqla-tester — COMMENTED** [review]
Hi, this is **sqla-tester** and I see you've pinged me for review. However, user **Daraan** is not authorized to initiate CI jobs.  Please wait for a project member to do this!

**Comment by CaselIT** [comment]
I've updated the change. I had to duplicate a bunch of tests sadly.
@zzzeek can you take a look? I've also added an utility to automate the "get type from both typing and typing_extension"

**Comment by Viicos** [comment]
Heads up from Pydantic: there is an edge case that need to be taken into account in Python 3.10. For more context, see [the definition of `typing_inspection.typing_objects.is_typealiastype()`](https://github.com/pydantic/typing-inspection/blob/e20451f3290a1e8833098749c766cb167ab83a40/src/typing_inspection/typing_objects.py#L477-L488).

A simplified implementation would look like:

```python
def is_typealiastype(obj):
    return type(obj) is not types.GenericAlias and isinstance(obj, (typing.TypeAliasType, typing_extensions.TypeAliasType))
```

**Comment by CaselIT** [comment]
when would that happen? if you use `[]` on a TypeAliasType?

**Comment by CaselIT** [comment]
> when would that happen? if you use `[]` on a TypeAliasType?

seems so. it wasn't taken into consideration at all the generic typealiastype case. should not be hard to handle, but I think for sqlalchemy we should consider the generic alias of tat a tat

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

* test/orm/declarative/test_typed_mapping.py (line 1298): this part is only on 2.0

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

the backport to 2.0 has some different test

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5825

**Comment by sqla-tester** [comment]
**Michael Bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5825

**Review by sqla-tester — COMMENTED** [review]
**Michael Bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

* test/orm/declarative/test_typed_mapping.py (line 1298): Done

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

it seems that 3.7 needs special handling

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

**Comment by Daraan** [comment]
> **Federico Caselli** (CaselIT) wrote:
> 
> it seems that 3.7 needs special handling
> 
> View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

I have a fix for at least one; can I commit it here or work with gerrit? If the later can you give me a quick rundown how to push/commit to it

**Comment by CaselIT** [comment]
no commits here will be ignored since the change in gerrit is quite a bit different.

It's likely easier if you post a diff here. Note that I was inclined to just skip that on 3.7. No one is using type alias types on it.

**Review by sqla-tester — COMMENTED** [review]
**Daraan** (Daraan) wrote:

Patch for the 3.7 tests as the TAT attributes are not availiable on a subscripted TATA (_GenericAlias)


```
diff --git a/test/base/test_typing_utils.py b/test/base/test_typing_utils.py
index bc83c5f2c..a9976c03a 100644
--- a/test/base/test_typing_utils.py
+++ b/test/base/test_typing_utils.py
@@ -425,21 +425,40 @@ class TestTyping(fixtures.TestBase):
         # generics
         eq_(sa_typing.pep695_values(TA_generic), {typing.List[TV]})
         eq_(sa_typing.pep695_values(TAext_generic), {typing.List[TV]})
-        eq_(sa_typing.pep695_values(TA_generic_typed), {typing.List[TV]})
-        eq_(sa_typing.pep695_values(TAext_generic_typed), {typing.List[TV]})
         eq_(sa_typing.pep695_values(TA_generic_null), {None, typing.List[TV]})
         eq_(
             sa_typing.pep695_values(TAext_generic_null),
             {None, typing.List[TV]},
         )
-        eq_(
-            sa_typing.pep695_values(TA_generic_null_typed),
-            {None, typing.List[TV]},
-        )
-        eq_(
-            sa_typing.pep695_values(TAext_generic_null_typed),
-            {None, typing.List[TV]},
-        )
+        if py38:
+            eq_(sa_typing.pep695_values(TA_generic_typed), {typing.List[TV]})
+            eq_(sa_typing.pep695_values(TAext_generic_typed), {typing.List[TV]})
+            eq_(
+                sa_typing.pep695_values(TAext_generic_null_typed),
+                {None, typing.List[TV]},
+            )
+            eq_(
+                sa_typing.pep695_values(TA_generic_null_typed),
+                {None, typing.List[TV]},
+            )
+        else:
+            # for 3.7 the attributes are only available on the origin
+            eq_(sa_typing.pep695_values(
+                typing_extensions.get_origin(TA_generic_typed)),
+                {typing.List[TV]}
+            )
+            eq_(sa_typing.pep695_values(
+                typing_extensions.get_origin(TAext_generic_typed)),
+                {typing.List[TV]}
+            )
+            eq_(sa_typing.pep695_values(
+                typing_extensions.get_origin(TAext_generic_null_typed)),
+                {None, typing.List[TV]}
+            )
+            eq_(
+                sa_typing.pep695_values(typing_extensions.get_origin(TA_generic_null_typed)),
+                {None, typing.List[TV]},
+            )
 
     def test_is_fwd_ref(self):
         eq_(sa_typing.is_fwd_ref(int), False)
@@ -526,9 +545,20 @@ class TestTyping(fixtures.TestBase):
         for t in type_aliases() + new_types():
             if t in false_negatives:
                 exp = False
-            else:
+            elif py38:
                 exp = "null" in t.__name__
-            eq_(sa_typing.includes_none(t), exp, str(t))
+                eq_(sa_typing.includes_none(t), exp, str(t))
+            else:
+                # on 3.7 the attributes are only available on the origin
+                origin = typing_extensions.get_origin(t)
+                if origin is None:
+                    # normal TypeAliasType
+                    exp = "null" in t.__name__
+                    eq_(sa_typing.includes_none(t), exp, str(t))
+                else:
+                    # subscripted TypeAliasType
+                    exp = "null" in origin.__name__
+                    eq_(sa_typing.includes_none(origin), exp, str(t))
 
         for t in annotated_l():
             eq_(
```

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

* test/base/test_typing_utils.py (line 428): fix for 3.7 tests:
the attributes are not forwarded to a subscribed TypeAliasType -> _GenericAlias

```
diff --git a/test/base/test_typing_utils.py b/test/base/test_typing_utils.py
index bc83c5f2c..a9976c03a 100644
--- a/test/base/test_typing_utils.py
+++ b/test/base/test_typing_utils.py
@@ -425,21 +425,40 @@ class TestTyping(fixtures.TestBase):
         # generics
         eq_(sa_typing.pep695_values(TA_generic), {typing.List[TV]})
         eq_(sa_typing.pep695_values(TAext_generic), {typing.List[TV]})
-        eq_(sa_typing.pep695_values(TA_generic_typed), {typing.List[TV]})
-        eq_(sa_typing.pep695_values(TAext_generic_typed), {typing.List[TV]})
         eq_(sa_typing.pep695_values(TA_generic_null), {None, typing.List[TV]})
         eq_(
             sa_typing.pep695_values(TAext_generic_null),
             {None, typing.List[TV]},
         )
-        eq_(
-            sa_typing.pep695_values(TA_generic_null_typed),
-            {None, typing.List[TV]},
-        )
-        eq_(
-            sa_typing.pep695_values(TAext_generic_null_typed),
-            {None, typing.List[TV]},
-        )
+        if py38:
+            eq_(sa_typing.pep695_values(TA_generic_typed), {typing.List[TV]})
+            eq_(sa_typing.pep695_values(TAext_generic_typed), {typing.List[TV]})
+            eq_(
+                sa_typing.pep695_values(TAext_generic_null_typed),
+                {None, typing.List[TV]},
+            )
+            eq_(
+                sa_typing.pep695_values(TA_generic_null_typed),
+                {None, typing.List[TV]},
+            )
+        else:
+            # for 3.7 the attributes are only available on the origin
+            eq_(sa_typing.pep695_values(
+                typing_extensions.get_origin(TA_generic_typed)),
+                {typing.List[TV]}
+            )
+            eq_(sa_typing.pep695_values(
+                typing_extensions.get_origin(TAext_generic_typed)),
+                {typing.List[TV]}
+            )
+            eq_(sa_typing.pep695_values(
+                typing_extensions.get_origin(TAext_generic_null_typed)),
+                {None, typing.List[TV]}
+            )
+            eq_(
+                sa_typing.pep695_values(typing_extensions.get_origin(TA_generic_null_typed)),
+                {None, typing.List[TV]},
+            )
```
* test/base/test_typing_utils.py (line 530): In 3.7 a subscripted typing_extensions.TypeAliasType is a rather simple typing._GenericAlias; it lacks a `__name__` field causing the tests to fail.

```python
get_origin(t).__name__
```

will work with 3.7:

patch:

```git
diff --git a/test/base/test_typing_utils.py b/test/base/test_typing_utils.py
index bc83c5f2c..04b3aa6c1 100644
--- a/test/base/test_typing_utils.py
+++ b/test/base/test_typing_utils.py
@@ -526,9 +545,20 @@ class TestTyping(fixtures.TestBase):
         for t in type_aliases() + new_types():
             if t in false_negatives:
                 exp = False
-            else:
+            elif py38:
                 exp = "null" in t.__name__
-            eq_(sa_typing.includes_none(t), exp, str(t))
+                eq_(sa_typing.includes_none(t), exp, str(t))
+            else:
+                # on 3.7 the attributes are only available on the origin
+                origin = typing_extensions.get_origin(t)
+                if origin is None:
+                    # normal TypeAliasType
+                    exp = "null" in t.__name__
+                    eq_(sa_typing.includes_none(t), exp, str(t))
+                else:
+                    # subscripted TypeAliasType
+                    exp = "null" in origin.__name__
+                    eq_(sa_typing.includes_none(origin), exp, str(t))
 
         for t in annotated_l():
             eq_(
```

**Comment by Daraan** [comment]
> no commits here will be ignored since the change in gerrit is quite a bit different.
> 
> It's likely easier if you post a diff here. Note that I was inclined to just skip that on 3.7. No one is using type alias types on it.

Posted a diff on gerrit; sadly I cannot edit there :/, hmm maybe on the original 5825?.

I had a bit trouble with disambiguating "no, commits here will be ignored" and "no commits will be ignored"😄, but I think I got it correctly that they are ignored.

**Comment by CaselIT** [comment]
I'm inclined to just skip those types in 37, since the lib code also seems to require an update, since it errors with 
> AttributeError: __value__

in
```
sqlalchemy/util/typing.py", line 387, in recursive_value
    value = inner_type.__value__
```

**Comment by Daraan** [comment]
> I'm inclined to just skip those types in 37, since the lib code also seems to require an update, since it errors with
> 
> > AttributeError: **value**
> 
> in
> 
> ```
> sqlalchemy/util/typing.py", line 387, in recursive_value
>     value = inner_type.__value__
> ```

Yes, makes more sense, and should be easier, to update the utility function for the py3.7 case :)

**Comment by CaselIT** [comment]
I think it's fine if we just tell people that use TAT on 3.7 to update the python version :)

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

* test/base/test_typing_utils.py (line 333): I'm not sure why this test has run even with this requires in the laterts pipeline

**Review by sqla-tester — COMMENTED** [review]
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

* test/base/test_typing_utils.py (line 333): I didnt test it b.c. i didnt have a python 3.7 here.   ill go onto the jenkins worker and investigate

**Review by sqla-tester — COMMENTED** [review]
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

* test/base/test_typing_utils.py (line 333): the git checkout on the worker failed in an entirely new and very ominous way.   going to see if it was a one off

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

* test/base/test_typing_utils.py (line 333): >  Took 7 sec on ryzen1
	
always it :(

**Review by sqla-tester — COMMENTED** [review]
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826

* test/base/test_typing_utils.py (line 428): the combination of these patches are failing and I'm not able to apply them, also I can't cut and paste from gerrit very easily since it does not allow me to scroll.   I am going to get the python38 decorators through first, if you want to create a complete patch for all fixes in a single file you can attach it to the PR or github issue

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5825 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5826 has been **merged**. Congratulations! :)

**Comment by Daraan** [comment]
A big thanks @ everyone for taking this over so fast.
